N-best candidates

sjarmero opened the pull request:
Returns the n-best candidates for a given input. The default value for n-best is 1. 

* Command line interface: n-best is set using the `-n` or `--nbest` flag. If unset, it defaults to 1.
* JoeyNMT Python API: `n_best` parameter can be provided in `translate` or `validate_on_data`.

sjarmero commented:
Added bug fixes and LINT fixes

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -488,7 +503,8 @@ def _translate_data(test_data):
                 hypotheses = _translate_data(test_data)
-                print("JoeyNMT: {}".format(hypotheses[0]))
+                for hyp in hypotheses:
+                    print("JoeyNMT: {}".format(hyp))
```
Perhaps indicate that these hypotheses are ranked by score (add an index?).

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -117,15 +122,23 @@ def validate_on_data(model: Model, data: Dataset,
                 attention_scores[sort_reverse_index]
                 if attention_scores is not None else [])
 
-        assert len(all_outputs) == len(data)
```
Replace with assertion of new length. Have you considered making the outputs a structured object rather than a flat list? As a flat list, it is not obvious which outputs correspond to which references (in the evaluation below), so they should rather be treated separately (i.e. separate evaluation for translations of rank0, rank1, etc, and written into separate files which all have the length of the references).

juliakreutzer reviewed (changes requested):
Thanks so much!! This is a very cool addition. I think we should treat the outputs slightly differently, see my comment.

sjarmero commented:
I have changed `all_outputs` and `sort_reverse_index` into a two-dimensional array, where the _n_ position stores outputs and indexes for the _n_ best hypothesis. I have also changed the code to output to different files if an `--output_path` is provided. If `n_best` equals 1, it will print to a single file. If `n_best` is greater than 1, it will print to `[file name]-[rank].[file extension]`

juliakreutzer commented on joeynmt/__main__.py:
```diff
@@ -32,7 +35,8 @@ def main():
         translate(cfg_file=args.config_path, ckpt=args.ckpt,
-                  output_path=args.output_path)
+                  output_path=args.output_path,
+                  n_best=args.nbest if args.nbest else 1)
```
Instead of making the check here, use 1 as default for the nbest argument. In this way the user can easily find out about the behavior of the flag in one place.

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -117,15 +125,17 @@ def validate_on_data(model: Model, data: Dataset,
+                    if attention_scores is not None else [])
 
-        assert len(all_outputs) == len(data)
+            assert len(all_outputs) == len(data)
```
This indentation seems too much. Is it intended?

sjarmero replied:
I've made the assertion to be checked in each loop. You're right, I can use the previous indent and run it just once.

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -459,18 +474,34 @@ def _translate_data(test_data):
+            if n_best > 1:
+                for n in range(n_best):
+                    file_name, file_extension = os.path.splitext(output_path)
+                    write_to_file(
```
Great, this is much better!

juliakreutzer reviewed (changes requested):
Looks good! Thanks for the improvements. Just one indent left, see comments :)

juliakreutzer reviewed (approved):
Looks good, thank you!

juliakreutzer commented:
@sjarmero please check out the issue found by@may-: #144. Could you please fix the two places in the code and make sure it runs? Thanks.
